batch.py

```diff
@@ -0,0 +1,761 @@
+    mt = mt.filter_rows(  # check these filters!
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)  # syntax???
+    logging.info(
+        f"Number of rare (freq<5%) and QC'd biallelic variants: {mt.count()[0]}"
+    )
+
+    return output_mt_path  # both input and output??
+
+
+# endregion SUBSET_VARIANTS
+
+
+# region GET_GENE_SPECIFIC_VARIANTS
+
+
+def get_promoter_variants(
+    mt_path: str,  # checkpoint from function above
+    ht_path: str,
+    gene_file: str,
+    gene_name: str,
+    window_size: int,
+    # plink_output_prefix: str, # figure out how to deal with this
+):
+    """Subset hail matrix table
+
+    Input:
+    mt_path: path to already subsetted hail matrix table
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) in promoter regions and
+    2) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    (on top of all filters done above)
+
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # get relevant chromosome
+    gene_df = pd.read_csv(AnyPath(dataset_path(gene_file)), sep="\t", index_col=0)
+    chrom = gene_df[gene_df["gene_name"] == gene_name]["chr"]
+    # subset to chromosome
+    mt = mt.filter_rows(mt.locus.contig == ("chr" + chrom))
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    interval_start = int(gene_df[gene_df["gene_name"] == gene_name]["start"]) - int(
+        window_size
+    )
+    interval_end = int(gene_df[gene_df["gene_name"] == gene_name]["end"]) + int(
+        window_size
+    )
+    # clip to chromosome boundaries
+    left_boundary = max(1, interval_start)
+    right_boundary = min(
+        interval_end, hl.get_reference("GRCh38").lengths[f"chr{chrom}"]
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"chr{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
```

Comment:
in hail terms it might be more expensive to do contig then region filtering as separate operations. `parse_locus_interval` already takes a contig, so you can safely remove the earlier contig filter

Reply:
@annacuomo  this would be an instance where passing the chr/start/end into this method directly could remove ~20 lines

Reply:
ah yes i think this was maybe also an issue (previously) with the densifying step? so since here i am passing an already densified mt i don't need this in two steps??

Reply:
AFAIK there shouldn't be efficiency gains, as (contig & position) form part of the row key so that should be an efficient search. Happy to be corrected!